feg/gateway/services/envoy_controller/control_plane/control_plane.go

```diff
@@ -0,0 +1,342 @@
+	var grpcOptions []grpc.ServerOption
+	grpcOptions = append(grpcOptions, grpc.MaxConcurrentStreams(grpcMaxConcurrentStreams))
+	grpcServer := grpc.NewServer(grpcOptions...)
+
+	lis, err := net.Listen("tcp", fmt.Sprintf(":%d", port))
+	if err != nil {
+		glog.Fatalf("failed to listen %s", err)
+	}
+
+	// register services
+	discovery.RegisterAggregatedDiscoveryServiceServer(grpcServer, server)
+	v2.RegisterEndpointDiscoveryServiceServer(grpcServer, server)
+	v2.RegisterClusterDiscoveryServiceServer(grpcServer, server)
+	v2.RegisterRouteDiscoveryServiceServer(grpcServer, server)
+	v2.RegisterListenerDiscoveryServiceServer(grpcServer, server)
+
+	glog.Infof("Management server listening on port %d", port)
+	go func() {
+		if err = grpcServer.Serve(lis); err != nil {
+			glog.Error(err)
+		}
+	}()
+	<-ctx.Done()
+
+	grpcServer.GracefulStop()
+}
+
+// RunManagementGateway starts an HTTP gateway to an xDS server.
+func RunManagementGateway(ctx context.Context, srv xds.Server, port uint) {
+	glog.Infof("Gateway listening HTTP/1.1 on port %d", port)
+	server := &http.Server{Addr: fmt.Sprintf(":%d", port), Handler: &xds.HTTPGateway{Server: srv}}
+	go func() {
+		if err := server.ListenAndServe(); err != nil {
+			glog.Error(err)
+		}
+	}()
+}
+
+func GetControllerClient() *ControllerClient {
+	cli := ControllerClient{}
+	ctx := context.Background()
+
+	glog.Infof("Starting Envoy control plane")
+
+	signal := make(chan struct{})
+	cb := &callbacks{
+		signal:   signal,
+		fetches:  0,
+		requests: 0,
+	}
+	cli.config = cache.NewSnapshotCache(mode == Ads, Hasher{}, nil)
+
+	srv := xds.NewServer(ctx, cli.config, cb)
+
+	// start the xDS server
+	go RunManagementServer(ctx, srv, port)
+	go RunManagementGateway(ctx, srv, gatewayPort)
+
+	<-signal
+
+	cb.Report()
+
+	return &cli
+}
+
+func (cli *ControllerClient) UpdateSnapshot(ues []*protos.AddUEHeaderEnrichmentRequest) {
+	nodeId := cli.config.GetStatusKeys()[0]
+
+	var clusterName = "cluster1"
+	cluster := []cache.Resource{
+		&v2.Cluster{
+			Name:                 clusterName,
+			ClusterDiscoveryType: &v2.Cluster_Type{Type: v2.Cluster_ORIGINAL_DST},
+			ConnectTimeout:       ptypes.DurationProto(6 * time.Second),
+			LbPolicy:             v2.Cluster_CLUSTER_PROVIDED,
+		},
+	}
+
+	filterChains := []*listener.FilterChain{}
+	var listenerName = "default_http"
+	var targetPrefix = "/"
+	var virtualHostName = "local_service"
+	var routeConfigName = "local_route"
+
+	glog.Errorf("Updating snapshot content")
+
+	for _, req := range ues {
+		var ue_ip_addr = string(req.UeIp.Address)
+		requestHeadersToAdd := []*core.HeaderValueOption{}
+
+		glog.Infof("Adding UE " + ue_ip_addr)
+
+		for _, header := range req.Headers {
+			requestHeadersToAdd = append(requestHeadersToAdd, &core.HeaderValueOption{
+				Header: &core.HeaderValue{
+					Key:   header.Name,
+					Value: header.Value,
+				},
+			})
+		}
+
+		virtualHosts := []*v2route.VirtualHost{}
+		virtualHosts = append(virtualHosts, &v2route.VirtualHost{
```

Comment:
how do we handle urls that do not match the req.Websites.

Reply:
good point, should we add a default case that doesn't add any headers?